test-case/check-alsabat.sh

```diff
@@ -50,39 +56,51 @@ then
 	exit 2
 fi
 
+pcmid_p=${pcm_p:-1}
+
 [[ ${OPT_VALUE_lst['s']} -eq 1 ]] && func_lib_start_log_collect
 
-function __upload_wav_file
-{
-    # upload the alsabat wav file
-    for file in /tmp/bat.wav.*
-    do
-	    size=`ls -l $file | awk '{print $5}'`
-	    if [[ $size -gt 0 ]]; then
-		    cp $file $LOG_ROOT/
-	    fi
-    done
-}
+func_pipeline_export "$tplg" "type:playback & id:$pcmid_p"
+
+# parser the parameters of the specified playback pipeline
+channel=$(func_pipeline_parse_value 0 ch_max)
+rate=$(func_pipeline_parse_value 0 rate)
+fmts=$(func_pipeline_parse_value 0 fmts)
 
 # check the PCMs before alsabat test
 dlogi "check the PCMs before alsabat test"
-[[ $(aplay -Dplug$pcm_p -d 1 /dev/zero -q) ]] && die "Failed to play on PCM: $pcm_p"
-[[ $(arecord -Dplug$pcm_c -d 1 /dev/null -q) ]] && die "Failed to capture on PCM: $pcm_c"
+[[ $(aplay -Dplug"$pcm_p" -d 1 /dev/zero -q) ]] && die "Failed to play on PCM: $pcm_p"
+[[ $(arecord -Dplug"$pcm_c" -d 1 /dev/null -q) ]] && die "Failed to capture on PCM: $pcm_c"
 
 # alsabat test
-# different PCMs may support different audio formats(like samplerate, channel-counting, etc.).
-# use plughw to do the audio format conversions. So we don't need to specify them for each PCM.
-dlogc "alsabat -Pplug$pcm_p --standalone -n $frames -F $frequency"
-alsabat -Pplug$pcm_p --standalone -n $frames -F $frequency &
-# playback may have low latency, add one second delay to aviod recording zero at beginning.
-sleep 1
-dlogc "alsabat -Cplug$pcm_c -F $frequency"
-alsabat -Cplug$pcm_c -F $frequency
-
-# upload failed wav file
-if [[ $? != 0 ]]; then
-	__upload_wav_file
-	exit 1
-fi
-
-exit 0
+for format in $fmts
+    do
+        if [ "$format" == "S24_LE" ]; then
+            dlogi "S24_LE is not supported, skip to test this format"
+            continue
+        fi
+
+        dlogc "alsabat -P $pcm_p --standalone -n $frames -F $frequency -c $channel -r $rate -f $format"
+        timeout -k 2 6 alsabat -P "$pcm_p" --standalone -n "$frames" -F "$frequency" -c "$channel" \
+		-r "$rate" -f "$format" & alsabatPID=$!
+        # playback may have low latency, add one second delay to aviod recording zero at beginning.
+        sleep 1
+
+        if func_nocodec_mode; then
+            format_c=$format
+            channel_c=$channel
+        else
+            # USB sound card only supports 1 channel S16_LE format.
+            format_c=S16_LE
+            channel_c=1
+        fi
+
+        dlogc "alsabat -C $pcm_c -F $frequency -f $format_c -c $channel_c -r $rate"
+        alsabat -C "$pcm_c" -F "$frequency" -f "$format_c" -c "$channel_c" -r "$rate" || {
+            func_upload_wav_file "/tmp" "bat.wav.*"
+            exit 1
+        }
+        wait $alsabatPID || die "alsabat -P failed to stop"
```

Comment:
It's a bit more complicated, sorry I should have spotted this the first time. To be correct this requires two different error messages. Something like this:
```shell
wait $alsabatPID && ret=$?
if [ $ret = 124 ]; then
   die "alsabat -P failed to stop - we timed it out"
elif [ $ret != 0 ] ; then
   die "alsabat -P failed, returned $ret"
fi
```
Untested code, please double-check in `man timeout` and test by this code by temporarily setting a very short timeout or wrong parameters.

Reply:
Right, if the command times out, and --preserve-status is not set, then exit with status 124.

Reply:
> Untested code, please double-check in man timeout and test by this code by temporarily setting a very short timeout or wrong parameters.

With a clearer, Monday morning brain I see the untested code I gave you was wrong: `ret` never gets initialized when `wait` fails because of short circuit evaluation. I'm afraid this means you didn't inject any error locally, didn't test it and never saw these messages printed, this is worrying.

Reply:
@marc-hb , I did performed the negative tests to verify this part of code, tried with wrong channel count "3" in alsabat command and returned 1.  if alsabat -P does not stop within the specified time, it will return 124. so which scenario you are talking above?

Reply:
I mean this:
```shell
ret=0
(exit 5) && ret=$?
echo $ret

=> 0
```

Reply:
I think we are overaction here. just put the alsabat -P in background and use wait to check the exit code, if exit code is non-zero， then exit with "alsabat -P failure".

Reply:
AFAIK @keqiaozhang didn't recommend to remove the `timeout` command, he only recommended to handle timeout errors exactly like non-timeout errors.

Question for people with `alsabat` experience: can `alsabat` get stuck  when something goes wrong?